pkg/controller/s3control/accesspoint/policy.go

```diff
@@ -0,0 +1,172 @@
+			if cr.Spec.ForProvider.Policy == nil {
+				return updated, nil
+			}
+			return needsUpdate, nil
+		}
+		return needsUpdate, aws.Wrap(err, errDescribePolicy)
+	}
+
+	if cr.Spec.ForProvider.Policy == nil {
+		if output.Policy != nil {
+			return needsDeletion, nil
+		}
+		return updated, nil
+	}
+
+	specPolicyRaw, err := formatPolicy(cr)
+	if err != nil {
+		return needsUpdate, errors.Wrap(err, errFormatPolicy)
+	}
+	specPolicy, err := policyutils.ParsePolicyString(aws.StringValue(specPolicyRaw))
+	if err != nil {
+		return needsUpdate, errors.Wrap(err, errParsePolicySpec)
+	}
+	curPolicy, err := policyutils.ParsePolicyString(aws.StringValue(output.Policy))
+	if err != nil {
+		return needsUpdate, errors.Wrap(err, errParseExternalPolicy)
+	}
+	diff := cmp.Diff(specPolicy, curPolicy)
+	if diff != "" {
+		return needsUpdate, nil
+	}
+	return updated, nil
+}
+
+// createOrUpdate creates or updates the Access Point policy.
+func (p *policyClient) createOrUpdate(ctx context.Context, cr *svcapitypes.AccessPoint) error {
+	if cr.Spec.ForProvider.Policy == nil {
+		return nil
+	}
+	input := &svcsdk.PutAccessPointPolicyInput{}
+	input.Name = aws.String(meta.GetExternalName(cr))
+
+	if cr.Spec.ForProvider.AccountID != nil {
+		input.SetAccountId(*cr.Spec.ForProvider.AccountID)
+	}
+
+	policy, err := formatPolicy(cr)
+	if err != nil {
+		return errors.Wrap(err, errFormatPolicy)
+	}
+	input.SetPolicy(*policy)
+	_, err = p.client.PutAccessPointPolicyWithContext(ctx, input)
+	return aws.Wrap(err, errPutPolicy)
+}
+
+// delete deletes the Access Point policy.
+func (p *policyClient) delete(ctx context.Context, cr *svcapitypes.AccessPoint) error {
+	_, err := p.client.DeleteAccessPointPolicyWithContext(ctx, generateDeleteAccessPointPolicyInput(cr))
+	return aws.Wrap(err, errDeletePolicy)
+}
+
+// isErrorPolicyNotFound returns whether the given error is of type NotFound or not.
+func isErrorPolicyNotFound(err error) bool {
+	var awsErr smithy.APIError
+	return errors.As(err, &awsErr) && awsErr.ErrorCode() == "NoSuchAccessPointPolicy"
+}
+
+// generateGetAccessPointPolicyInput returns a get input.
+func generateGetAccessPointPolicyInput(cr *svcapitypes.AccessPoint) *svcsdk.GetAccessPointPolicyInput {
+	input := &svcsdk.GetAccessPointPolicyInput{}
+	input.Name = aws.String(meta.GetExternalName(cr))
+
+	if cr.Spec.ForProvider.AccountID != nil {
+		input.SetAccountId(*cr.Spec.ForProvider.AccountID)
+	}
+
+	return input
+}
+
+// generateDeleteAccessPointPolicyInput returns a deletion input.
+func generateDeleteAccessPointPolicyInput(cr *svcapitypes.AccessPoint) *svcsdk.DeleteAccessPointPolicyInput {
+	input := &svcsdk.DeleteAccessPointPolicyInput{}
+	input.Name = aws.String(meta.GetExternalName(cr))
+	if cr.Spec.ForProvider.AccountID != nil {
+		input.SetAccountId(*cr.Spec.ForProvider.AccountID)
+	}
+
+	return input
+}
+
+// formatPolicy parses and formats the Policy struct
+func formatPolicy(cr *svcapitypes.AccessPoint) (*string, error) {
```

Comment:
As said above: It is better to reuse the implementation in `clients/s3` rather than having duplicated code.

Reply:
Same question as above, Should I make a common function for formatting policy or simply use `s3.Serialize` inside `formatPolicy` func?

Reply:
Yes, a common logic would be best.